File: src/main/java/net/reflact/common/attribute/AttributeContainer.java
package net.reflact.common.attribute;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class AttributeContainer {
    private final Map<String, List<AttributeModifier>> modifiers = new ConcurrentHashMap<>();

    public void addModifier(Attribute attribute, AttributeModifier modifier) {
        modifiers.computeIfAbsent(attribute.id(), k -> new ArrayList<>()).add(modifier);
    }

    public void removeModifier(Attribute attribute, String modifierId) {
        if (modifiers.containsKey(attribute.id())) {
            modifiers.get(attribute.id()).removeIf(mod -> mod.getId().equals(modifierId));
        }
    }

    public void clearModifiersByPrefix(String prefix) {
        for (List<AttributeModifier> mods : modifiers.values()) {
            mods.removeIf(mod -> mod.getId().startsWith(prefix));
        }
    }

    public double getValue(Attribute attribute) {
        if (attribute == null) return 0;
        
        double base = attribute.baseValue();
        List<AttributeModifier> mods = modifiers.getOrDefault(attribute.id(), Collections.emptyList());

        double addNumber = 0;
        double addScalar = 0;
        double multiplyScalar = 1;

        for (AttributeModifier mod : mods) {
            switch (mod.getOperation()) {
                case ADD_NUMBER -> addNumber += mod.getValue();
                case ADD_SCALAR -> addScalar += mod.getValue();
                case MULTIPLY_SCALAR -> multiplyScalar *= mod.getValue();
            }
        }

        double value = (base + addNumber) * (1 + addScalar) * multiplyScalar;
        return Math.min(Math.max(value, attribute.minValue()), attribute.maxValue());
    }
}

File: bin/main/net/reflact/common/attribute/AttributeContainer.kt
package net.reflact.common.attribute

import java.util.concurrent.ConcurrentHashMap

class AttributeContainer {
    private val modifiers: MutableMap<String, MutableList<AttributeModifier>> = ConcurrentHashMap()

    fun addModifier(attribute: Attribute, modifier: AttributeModifier) {
        modifiers.computeIfAbsent(attribute.id) { ArrayList() }.add(modifier)
    }

    fun removeModifier(attribute: Attribute, modifierId: String) {
        if (modifiers.containsKey(attribute.id)) {
            modifiers[attribute.id]?.removeIf { it.id == modifierId }
        }
    }

    fun clearModifiersByPrefix(prefix: String) {
        for (mods in modifiers.values) {
            mods.removeIf { it.id.startsWith(prefix) }
        }
    }

    fun getValue(attribute: Attribute?): Double {
        if (attribute == null) return 0.0

        val base = attribute.baseValue
        val mods = modifiers.getOrDefault(attribute.id, emptyList())

        var addNumber = 0.0
        var addScalar = 0.0
        var multiplyScalar = 1.0

        for (mod in mods) {
            when (mod.operation) {
                AttributeModifier.Operation.ADD_NUMBER -> addNumber += mod.value
                AttributeModifier.Operation.ADD_SCALAR -> addScalar += mod.value
                AttributeModifier.Operation.MULTIPLY_SCALAR -> multiplyScalar *= mod.value
            }
        }

        val value = (base + addNumber) * (1 + addScalar) * multiplyScalar
        return value.coerceIn(attribute.minValue, attribute.maxValue)
    }
}
